### PR TITLE
feat: add optimize command

### LINKED ISSUE #218: Add `asyncapi optimize` command
> This issue is for https://www.openforce.tech/ participants. If you do not participate in this even please pick up other issue from this repo, there is plenty of help needed 😄 

#### Reason/Context

We need to enable the community to use all AsyncAPI tools in CLI if only possible -> https://github.com/asyncapi/optimizer

#### Description

- add new command `asyncapi optimize` that uses https://github.com/asyncapi/optimizer and enable users to optimize single AsyncAPI file
- user should be able to pass one files to the optimizer `asyncapi optimize <context/filepath/url>` (getting file from context or filepath or URL is already provided in  CLI, you just use `load` function)
- by default it would optimize the document and print to the terminal, but user should be able to specify to get only the report, with flag `--report`
- by default full optimization is done, but user should be able to specify `--rules` where user can provide rules that should be applied by optimizer

> This issue is for https://www.openforce.tech/ participants. If you do not participate in this even please pick up other issue from this repo, there is plenty of help needed 😄 

@KhudaDad414 pinging you so you are aware 😄

### PR DESCRIPTION
**Pull request by KhudaDad414**
**Description**

- Integrate asyncapi/optimizer as `asyncapi optimize` command.

**Related issue(s)**
Fixes #218 , to be merged after https://github.com/asyncapi/optimizer/pull/129

### CODE REVIEW

**Review by derberg — CHANGES_REQUESTED** [review]
- please update readme
- please add different examples for command
- also after checking out repo, installing and running build I get below. Not sure what I'm doing wrong, it should work otherwise how would you be able to play with it 🤔 
```
src/commands/optimize.ts:2:21 - error TS2305: Module '"@asyncapi/optimizer"' has no exported member 'Output'.

2 import { Optimizer, Output, Report, ReportElement } from '@asyncapi/optimizer';
                      ~~~~~~

src/commands/optimize.ts:2:29 - error TS2305: Module '"@asyncapi/optimizer"' has no exported member 'Report'.

2 import { Optimizer, Output, Report, ReportElement } from '@asyncapi/optimizer';
                              ~~~~~~

src/commands/optimize.ts:2:37 - error TS2305: Module '"@asyncapi/optimizer"' has no exported member 'ReportElement'.

2 import { Optimizer, Output, Report, ReportElement } from '@asyncapi/optimizer';
                                      ~~~~~~~~~~~~~

src/commands/optimize.ts:50:11 - error TS2532: Object is possibly 'undefined'.

50     if (!(report.moveToComponents.length || report.removeComponents.length || report.reuseComponents.length)) {
             ~~~~~~~~~~~~~~~~~~~~~~~

src/commands/optimize.ts:50:45 - error TS2532: Object is possibly 'undefined'.

50     if (!(report.moveToComponents.length || report.removeComponents.length || report.reuseComponents.length)) {
                                               ~~~~~~~~~~~~~~~~~~~~~~~

src/commands/optimize.ts:50:79 - error TS2532: Object is possibly 'undefined'.

50     if (!(report.moveToComponents.length || report.removeComponents.length || report.reuseComponents.length)) {
                                                                                 ~~~~~~~~~~~~~~~~~~~~~~

src/commands/optimize.ts:116:57 - error TS7006: Parameter 'e' implicitly has an 'any' type.

116       const totalMove = report.moveToComponents.filter((e) => e.action === 'move').length;
                                                            ~


Found 7 errors in the same file, starting at: src/commands/optimize.ts:2
```

**Comment by KhudaDad414** [comment]
@derberg @Souvikns a review would be appreciated 🙏

**Review by Souvikns — APPROVED** [review]
LGTM, @KhudaDad414 you should still wait for @derberg review.

File: src/commands/optimize.ts
Comment (derberg) [inline src/commands/optimize.ts:85]: please make sure we generate file called `optimized-asyncapi.yml`

we are trying to make sure autocompletion is supported for IDEs and we added AsyncAPI schema to https://github.com/smoya/schemastore/blob/master/src/api/json/catalog.json#L239-L249. There is some filename matching that enables autocompletion by default.

yes autocompletion do not really work in IDEs yet (soon https://github.com/asyncapi/spec-json-schemas/pull/298) but will 🚀 and we should make sure we promote proper naming of files.

File: src/commands/optimize.ts
Comment (derberg) [inline src/commands/optimize.ts:55]: optimizer uses parser internally, so indirectly it also validates asyncapi documents. If this validation fails then `Error: There were errors validating the AsyncAPI document.` is thrown instead of `ValidationError`. I'm not 100% sure as when I look at your code I think you throw error properly.

we definitely need separate test coverage for it

File: src/commands/optimize.ts
Comment (derberg) [inline src/commands/optimize.ts:62]: also resulting file should be specified, its location, so from console link you can quickly access it, especially when calling `asyncapi optimize ./asyncapi.yaml --no-tty`

**Review by derberg — CHANGES_REQUESTED** [review]
yo, left some comments

btw, played locally with interactiveness, nifty!!!

File: src/commands/optimize.ts
Comment (derberg) [inline src/commands/optimize.ts:50]: in other places we do

```
    try {
      specification = await load(asyncapi);
    } catch (err: any) {
      return this.error(
        new ValidationError({
          type: 'invalid-file',
          filepath: asyncapi,
        }),
        { exit: 1 },
      );
    }
```

File: test/commands/generate/fromTemplate.test.ts
Comment (KhudaDad414) [reply test/commands/generate/fromTemplate.test.ts:28]: @Souvikns and I have been tinkering with this test a lot lately. the thing is that this test fails randomly. in local and in [CI/CD pipeline](https://github.com/asyncapi/cli/actions/runs/3693882406/jobs/6254400684). We thought it would be better to get rid of it.

File: src/commands/optimize.ts
Comment (KhudaDad414) [reply src/commands/optimize.ts:55]: Since the asyncapi file is being validated before reaching the optimizer once, I don't see why it should fail when the optimizer parses it. I am not exactly sure how should I test it. 🤔

File: src/commands/optimize.ts
Comment (KhudaDad414) [reply src/commands/optimize.ts:62]: this message is shown when the file can't be optimized since there are no optimizations to be applied. ( I will make the message a little clearer.)

File: src/commands/optimize.ts
Comment (derberg) [reply src/commands/optimize.ts:55]: > I don't see why it should fail when the optimizer parses it

I didn't get it

Comment (KhudaDad414) [reply src/commands/optimize.ts:55]: sorry, my bad, I thought the asyncapi file is parsed by cli itself before getting to the optimizer. working on it right now. don't mind that comment.

File: test/commands/generate/fromTemplate.test.ts
Comment (derberg) [reply test/commands/generate/fromTemplate.test.ts:28]: instead of removing, in separate PR probably, we need just a fix for test. I remember having lots of random fails on `diff` command tests and we had to rework these to have each test in separate `describe`.

in short https://github.com/asyncapi/cli/blob/master/test/commands/generate/fromTemplate.test.ts#L17-L56 -> all 3 tests should be separated in individual `describe` as they probably overlap each other because of some async shit 😄

**Comment by KhudaDad414** [comment]
@derberg @Souvikns another round of 👁️ 🙏

**Comment by Souvikns** [comment]
I think you should resolve the conflict first, it will dismiss the review later.

**Comment by derberg** [comment]
@KhudaDad414 the tests are red

**Comment by KhudaDad414** [comment]
@derberg the tests are red all over this repo. As you can see in the commit history, I have bashed my head against the wall tooooo many times over these failing tests and achieved nothing. 😆

**Comment by derberg** [comment]
@KhudaDad414 please update to latest master, there are some conflicts

**Comment by KhudaDad414** [comment]
@derberg I think it is ready.

**Comment by derberg** [comment]
@KhudaDad414 relax, zen, and all that stuff..... conflicts again....so changes in docs in readme. You need to pull latest and move info about new command to respective markdown file in `docs` directory 😄

File: docs/usage.md
Comment (KhudaDad414) [inline docs/usage.md:24]:
```suggestion
 optimizes your AsyncAPI file by removing unused components, re-using duplicates etc.
```

**Review by derberg — APPROVED** [review]
awesome work and amazing patience 😆

**Review by Souvikns — APPROVED** [review]
👍🏼

**Pull request merged**
